File: yasmin_ros/test/test_monitor_state.cpp
// Copyright (C) 2025 Miguel Ángel González Santamarta
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

#include <chrono>
#include <gtest/gtest.h>
#include <memory>
#include <thread>

#include "std_msgs/msg/string.hpp"

#include "rclcpp/rclcpp.hpp"

#include "yasmin_ros/action_state.hpp"
#include "yasmin_ros/basic_outcomes.hpp"
#include "yasmin_ros/monitor_state.hpp"
#include "yasmin_ros/ros_clients_cache.hpp"
#include "yasmin_ros/ros_logs.hpp"
#include "yasmin_ros/yasmin_node.hpp"

using namespace yasmin_ros;
using namespace yasmin_ros::basic_outcomes;
using namespace std::chrono_literals;

class AuxNode : public rclcpp::Node {
public:
  AuxNode() : Node("test_node") {
    // Publisher
    this->publisher_ =
        this->create_publisher<std_msgs::msg::String>("test", 10);
    this->timer_ =
        this->create_wall_timer(1s, std::bind(&AuxNode::publish_message, this));
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;

  void publish_message() {
    if (rclcpp::ok()) {
      auto message = std_msgs::msg::String();
      message.data = "data";
      publisher_->publish(message);
    }
  }
};

class TestMonitorState : public ::testing::Test {
protected:
  static std::shared_ptr<AuxNode> aux_node;
  static std::shared_ptr<rclcpp::executors::MultiThreadedExecutor> executor;
  static std::thread spin_thread;

  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
    aux_node = std::make_shared<AuxNode>();
    executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
    executor->add_node(aux_node);
    spin_thread = std::thread([&]() { executor->spin(); });
  }

  static void TearDownTestSuite() {
    executor->cancel();
    if (spin_thread.joinable()) {
      spin_thread.join();
    }
    ROSClientsCache::clear_all();
    aux_node.reset();
    executor.reset();
    rclcpp::shutdown();
  }
};

std::shared_ptr<AuxNode> TestMonitorState::aux_node = nullptr;
std::shared_ptr<rclcpp::executors::MultiThreadedExecutor>
    TestMonitorState::executor = nullptr;
std::thread TestMonitorState::spin_thread;

TEST_F(TestMonitorState, TestMonitorTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state = std::make_shared<MonitorState<std_msgs::msg::String>>(
      "test1", std::set<std::string>{std::string(SUCCEED)},
      [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
         std::shared_ptr<std_msgs::msg::String> msg) {
        return std::string(SUCCEED);
      },
      rclcpp::QoS(10), 10, 2);

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

TEST_F(TestMonitorState, TestMonitorRetryTimeout) {
  auto blackboard = std::make_shared<yasmin::blackboard::Blackboard>();

  auto state = std::make_shared<MonitorState<std_msgs::msg::String>>(
      "test1", std::set<std::string>{std::string(SUCCEED)},
      [](std::shared_ptr<yasmin::blackboard::Blackboard> blackboard,
         std::shared_ptr<std_msgs::msg::String> msg) {
        return std::string(SUCCEED);
      },
      rclcpp::QoS(10), 10, 2, 3); // timeout=2, maximum_retry=3

  EXPECT_EQ((*state)(blackboard), std::string(TIMEOUT));
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: yasmin_ros/test/test_monitor_state.py
# Copyright (C) 2023 Miguel Ángel González Santamarta
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import unittest
from threading import Thread

from yasmin import set_py_loggers
from yasmin_ros import MonitorState
from yasmin_ros.basic_outcomes import SUCCEED, TIMEOUT
from yasmin_ros.yasmin_node import YasminNode
from yasmin_ros.ros_clients_cache import ROSClientsCache

import rclpy
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from std_msgs.msg import String


class AuxNode(Node):
    def __init__(self):
        super().__init__("test_node")

        self.pub = self.create_publisher(String, "test", 10)
        self.timer = self.create_timer(1, self.publish_msgs)

    def publish_msgs(self) -> None:
        msg = String()
        msg.data = "data"
        self.pub.publish(msg)


class TestYasminRos(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        rclpy.init()
        set_py_loggers()

        cls.aux_node = AuxNode()
        cls.executor = MultiThreadedExecutor()
        cls.executor.add_node(cls.aux_node)

        cls.spin_thread = Thread(target=cls.executor.spin)
        cls.spin_thread.start()

    @classmethod
    def tearDownClass(cls):
        ROSClientsCache.clear_all()
        YasminNode.destroy_instance()
        rclpy.shutdown()

    def test_monitor_timeout(self):

        def monitor_handler(blackboard, msg):
            return SUCCEED

        state = MonitorState(
            String,
            "test1",
            [SUCCEED],
            monitor_handler=monitor_handler,
            timeout=2,
        )
        self.assertEqual(TIMEOUT, state())

    def test_monitor_retry_timeout(self):
        def monitor_handler(blackboard, msg):
            return SUCCEED

        retries = 3

        ## Capture the logs
        with self.assertLogs(level="WARNING") as captured:
            state = MonitorState(
                String,
                "test1",
                [SUCCEED],
                monitor_handler=monitor_handler,
                timeout=0.1,
                maximum_retry=retries,
            )
            self.assertEqual(TIMEOUT, state())

        ## Check that the number of WARNING logs is correct
        self.assertEqual(
            retries,
            len(captured.records),
            msg=f"Expected {retries} WARNING logs, saw {len(captured)}.\n"
            f"Captured messages: {[r.getMessage() for r in captured.records]}",
        )


if __name__ == "__main__":
    unittest.main()
